pkg/controllers/provisioning/v1alpha1/reallocation/terminate.go

```diff
@@ -19,21 +19,125 @@ import (
+	if err := t.drainAndDeleteNodes(ctx, provisioner); err != nil {
+		return fmt.Errorf("draining nodes, %w", err)
+	}
+	return nil
+}
+
+// cordonNodes takes in a list of expired nodes as input and cordons them
+func (t *Terminator) cordonNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	nodeList, err := t.getNodes(ctx, provisioner, map[string]string{})
+	if err != nil {
+		return err
+	}
+
+	nodes := []*v1.Node{}
+	for _, node := range nodeList {
+		if utilsnode.IsPastTTL(node) && !node.Spec.Unschedulable {
+			nodes = append(nodes, node)
+		}
+	}
+
+	for _, node := range nodes {
+		persisted := node.DeepCopy()
+		node.Spec.Unschedulable = true
+		node.Labels[v1alpha1.ProvisionerDrainingKey] = "true"
+		if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			return fmt.Errorf("patching node %s, %w", node.Name, err)
+		}
+		zap.S().Debugf("Cordoned node %s", node.Name)
+	}
+	return nil
+}
+
+// drainAndDeleteNodes takes in a list of expired non-drained nodes and calls drain on them
+func (t *Terminator) drainAndDeleteNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) error {
+	drained := []*v1.Node{}
+	draining, err := t.getNodes(ctx, provisioner, map[string]string{
+		v1alpha1.ProvisionerDrainingKey: "true",
+	})
+	if err != nil {
+		return fmt.Errorf("listing draining nodes, %w", err)
+	}
+
+	for _, node := range draining {
+		// (0.) TODO: Check if Node should be drained
+		// - Disrupts PDB
+		// - Pods owned by controller object
+		// - Pod on Node can't be rescheduled elsewhere
+
+		// 1. Get Pods on Node
+		podList := &v1.PodList{}
+		if err := t.kubeClient.List(ctx, podList, client.MatchingFields{"spec.nodeName": node.Name}); err != nil {
+			zap.S().Debugf("Continuing after failing to list pods for node %s, %s", node.Name, err.Error())
+		}
+		// 2. If no pods on node, label as drained
+		empty := true
+		for _, pod := range ptr.PodListToSlice(podList) {
+			if !scheduling.IsOwnedByDaemonSet(pod) {
+				empty = false
+				if err := t.coreV1Client.Pods(pod.Namespace).Evict(ctx, &v1beta1.Eviction{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: pod.Name,
+					},
+				}); err != nil {
+					zap.S().Debugf("Continuing after failing to evict pods from node %s, %s", node.Name, err.Error())
+				}
+			}
+		}
+
+		// 3. Patch Node label if it is draining
+		persisted := node.DeepCopy()
+		if !empty {
+			// If not empty, keep it as draining
+			node.Labels[v1alpha1.ProvisionerDrainingKey] = "true"
+		} else {
+			// If empty, mark as drained
+			drained = append(drained, node)
+		}
+		if err := t.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			zap.S().Debugf("Continuing after failing to patch node %s, %s", node.Name, err.Error())
+		}
+	}
+
+	// 4. Delete any nodes that were fully drained
+	if err := t.deleteNodes(ctx, drained, provisioner); err != nil {
+		return fmt.Errorf("deleting %d nodes, %w", len(drained))
```

Comment:
This is the one place where I think we need to have custom error handling - a node can fail to terminate due to some kind of hang up on the control plane side (say something got attached to it we're not aware of, etc) and we should still terminate the other nodes. A simple solution would be to randomize the order that we traverse the nodes to terminate (this is my recommendation), but we could also get into handling different error conditions (differentiate between the API server just throttling us or some kind of malformed input vs. this-instance-cannot-be-terminated-because-reason).

Reply:
I'd be fine w/ it happening in the cloud provider.